include/mxnet/engine.h

```diff
@@ -74,15 +74,15 @@ class CallbackOnComplete {
  public:
   // use implicit copy and assign
   /*! \brief involve the callback */
-  inline void operator()() const {
-    (*callback_)(engine_, param_);
+  inline void operator()(const char* error_msg = nullptr) const {
```

Comment:
Instead of passing in a const string of error message, would it make sense to pass in a `error` struct which contains error message and error code in the fields? This leaves more room for proper error handling based on error code.

Reply:
The error code is not universally defined across different libraries. In the Horovod case, the error types are mostly Horovod specific. We convert all those to dmlc::Error which MXNet can catch.

Reply:
Can we define an error data structure and export it? So all libraries that consume libmxnet will use this error data structure to pass data back.

Reply:
This sounds like a good idea to make it extendable in the future. We can use the error message for now.

Reply:
sgtm

Reply:
Use dmlc::Error instead of char*. We can change dmlc::Error if more information need to be passed.